rust/connlib/libs/tunnel/src/tun_android.rs

```diff
@@ -1,20 +1,87 @@
 use super::InterfaceConfig;
-use libs_common::Result;
+use ip_network::IpNetwork;
+use libc::{close, open, O_RDWR};
+use libs_common::{Error, Result};
+use std::{
+    os::fd::{AsRawFd, RawFd},
+    sync::Arc,
+};
 
 #[derive(Debug)]
 pub(crate) struct IfaceConfig(pub(crate) Arc<IfaceDevice>);
 
 #[derive(Debug)]
-pub(crate) struct IfaceDevice;
+pub(crate) struct IfaceDevice {
+    fd: RawFd,
+}
+
+impl AsRawFd for IfaceDevice {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+impl Drop for IfaceDevice {
+    fn drop(&mut self) {
+        unsafe { close(self.fd) };
+    }
+}
+
+impl IfaceDevice {
+    fn write(&self, _buf: &[u8]) -> usize {
+        tracing::error!("`write` unimplemented on Android");
+        0
+    }
+
+    pub async fn new(_name: &str) -> Result<Self> {
+        // TODO: This won't actually work for non-root users...
+        let fd = unsafe { open(b"/dev/net/tun\0".as_ptr() as _, O_RDWR) };
+        // TODO: everything!
+        if fd == -1 {
+            Err(Error::Io(std::io::Error::last_os_error()))
+        } else {
+            Ok(Self { fd })
+        }
```

Comment:
It is 100% preference so feel free to discard but I find that using early return and avoiding else to be much easier to read:

```suggestion
        if fd == -1 {
            return Err(Error::Io(std::io::Error::last_os_error()))
        }
        
        Ok(Self { fd })
```

Reply:
I empathize with you viewpoint, but I consider early return to be a larger hammer, and branching has the additional benefit of making it easier to visualize potential combinator golfs/etc.